std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
+    });
+    const semicolon_token = try expectToken(it, tree, .Semicolon);
+    const use_node_raw = use_node.cast(Node.Use).?;
+    use_node_raw.*.doc_comments = doc_comments;
+    use_node_raw.*.visib_token = vis;
+    use_node_raw.*.expr = expr_node;
+    use_node_raw.*.semicolon_token = semicolon_token;
+
+    return use_node;
+}
 
-            State.FnDef => |fn_proto| {
-                const token = nextToken(&tok_it, tree);
-                const token_index = token.index;
-                const token_ptr = token.ptr;
-                switch (token_ptr.id) {
-                    Token.Id.LBrace => {
-                        const block = try arena.create(ast.Node.Block);
-                        block.* = ast.Node.Block{
-                            .base = ast.Node{ .id = ast.Node.Id.Block },
-                            .label = null,
-                            .lbrace = token_index,
-                            .statements = ast.Node.Block.StatementList.init(arena),
-                            .rbrace = undefined,
-                        };
-                        fn_proto.body_node = &block.base;
-                        stack.append(State{ .Block = block }) catch unreachable;
-                        continue;
-                    },
-                    Token.Id.Semicolon => continue,
-                    else => {
-                        ((try tree.errors.addOne())).* = Error{ .ExpectedSemiOrLBrace = Error.ExpectedSemiOrLBrace{ .token = token_index } };
-                        return tree;
-                    },
-                }
-            },
-            State.FnProto => |fn_proto| {
-                stack.append(State{ .FnProtoAlign = fn_proto }) catch unreachable;
-                try stack.append(State{ .ParamDecl = fn_proto });
-                try stack.append(State{ .ExpectToken = Token.Id.LParen });
+// FnProto <- FnCC? KEYWORD_fn IDENTIFIER? LPAREN ParamDeclList RPAREN ByteAlign? LinkSection? EXCLAMATIONMARK? (KEYWORD_var / TypeExpr)
+fn parseFnProto(arena: *Allocator, it: *TokenIterator, tree: *Tree) !?*Node {
+    const cc = try parseFnCC(arena, it, tree);
+    const fn_token = eatToken(it, .Keyword_fn) orelse {
+        if (cc == null) return null else return Error.UnexpectedToken;
+    };
+    const name_token = eatToken(it, .Identifier);
+    const lparen = try expectToken(it, tree, .LParen);
+    const params = try parseParamDeclList(arena, it, tree);
+    const rparen = try expectToken(it, tree, .RParen);
+    const align_expr = try parseByteAlign(arena, it, tree);
+    const section_expr = try parseLinkSection(arena, it, tree);
+    const exclamation_token = eatToken(it, .Bang);
+
+    const return_type_expr = (try parseVarType(arena, it, tree)) orelse
+        try expectNode(arena, it, tree, parseTypeExpr, AstError{
+        .ExpectedReturnType = AstError.ExpectedReturnType{ .token = it.index },
+    });
+
+    // TODO: Based on this rule, `!var` is an acceptable return type, but there is no usage
```

Comment:
I see now, this is https://github.com/ziglang/zig/issues/447 (Note that that proposal is marked accepted.) For some reason I thought this was an existing feature.

I'll extract this TODO grammar idea and just leave a comment on that ticket in case it's useful.